bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/ICalendarHandler.java

```diff
@@ -0,0 +1,396 @@
+                            new DateTimeType(currentEvent.end.atZone(ZoneId.systemDefault())));
+                }
+            } else {
+                updateState(eventPresenceChannel.getUID(), OnOffType.OFF);
+                updateState(currentEventTitleChannel.getUID(), UnDefType.NULL);
+                updateState(currentEventStartChannel.getUID(), UnDefType.NULL);
+                updateState(currentEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            Event nextEvent = calendar.getNextEvent(now);
+            if (nextEvent != null) {
+                updateState(nextEventTitleChannel.getUID(), new StringType(nextEvent.title));
+                updateState(nextEventStartChannel.getUID(),
+                        new DateTimeType(nextEvent.start.atZone(ZoneId.systemDefault())));
+                updateState(nextEventEndChannel.getUID(),
+                        new DateTimeType(nextEvent.end.atZone(ZoneId.systemDefault())));
+            } else {
+                updateState(nextEventTitleChannel.getUID(), UnDefType.NULL);
+                updateState(nextEventStartChannel.getUID(), UnDefType.NULL);
+                updateState(nextEventEndChannel.getUID(), UnDefType.NULL);
+            }
+
+            // process all Command Tags in all Calendar Events which ENDED since updateStates was last called
+            // the END Event tags must be processed before the BEGIN ones
+            executeEventCommands(calendar.getJustEndedEvents(updateStatesLastCalledTime, now), CommandTagType.END);
+
+            // process all Command Tags in all Calendar Events which BEGAN since updateStates was last called
+            // the END Event tags must be processed before the BEGIN ones
+            executeEventCommands(calendar.getJustBegunEvents(updateStatesLastCalledTime, now), CommandTagType.BEGIN);
+
+            // save time when updateStates was previously called
+            // the purpose is to prevent repeat command execution of events that have already been executed
+            updateStatesLastCalledTime = now;
+        }
+    }
+
+    private void executeEventCommands(@Nullable List<Event> events, CommandTagType execTime) {
+        // no begun or ended events => exit quietly as there is nothing to do
+        if ((events == null) || (events.isEmpty())) {
+            return;
+        }
+        // prevent synchronization issues (MVN null pointer warnings) in "eventPublisherCallback"
+        @Nullable
+        EventPublisher syncEventPublisherCallback = eventPublisherCallback;
+        if (syncEventPublisherCallback == null) {
+            logger.error("EventPublisher object not instantiated!");
+            return;
+        }
+        // prevent potential synchronization issues (MVN null pointer warnings) in "configuration"
+        @Nullable
+        ICalendarConfiguration syncConfiguration = configuration;
+        if (syncConfiguration == null) {
+            logger.error("Configuration not instantiated!");
```

Comment:
This should probably be a DEBUG, but can't be above WARN.

Reply:
Same as above.